File: scripts/rust/src/review_stt_compliance.py
#!/usr/bin/env python3
"""
Detect which data structures don't implement Clone + Display + Debug (StT requirements).

StT (Single-Threaded Type) = Eq + Clone + Display + Debug + Sized
StTInMtT = StT + Send + Sync
MtT (Multi-Threaded Type) = Sized + Send + Sync + has Inner: StT

This script finds public structs that DON'T satisfy StT requirements.
"""
# Git commit: 584a672b6a34782766863c5f76a461d3297a741a
# Date: 2025-10-17 05:17:36 -0700


import re
import sys
from pathlib import Path
from collections import defaultdict

# Add lib directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'lib'))
from review_utils import ReviewContext, create_review_parser


def extract_derives(lines, struct_line_idx):
    """Extract derives from the lines before a struct definition."""
    derives = set()
    
    # Look backwards for #[derive(...)]
    for i in range(struct_line_idx - 1, max(0, struct_line_idx - 5), -1):
        line = lines[i].strip()
        if line.startswith('#[derive('):
            # Extract traits from derive: #[derive(Debug, Clone, Copy, ...)]
            match = re.search(r'#\[derive\((.*?)\)\]', line)
            if match:
                traits_str = match.group(1)
                for trait in traits_str.split(','):
                    trait = trait.strip()
                    derives.add(trait)
        elif not line.startswith('#'):
            # Stop at non-attribute line
            break
    
    return derives


def has_manual_impl(lines, struct_name, trait_name):
    """Check if struct has a manual impl for trait."""
    # Look for: impl ... Trait for StructName
    pattern = rf'impl.*{trait_name}\s+for\s+{struct_name}'
    for line in lines:
        if re.search(pattern, line):
            return True
    return False


def analyze_file(filepath, context):
    """Analyze a file for structs that don't satisfy StT."""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            lines = f.readlines()
    except Exception:
        return []
    
    non_stt_structs = []
    
    for i, line in enumerate(lines):
        # Find pub struct definitions
        match = re.match(r'\s*pub\s+struct\s+(\w+)', line)
        if match:
            struct_name = match.group(1)
            
            # Extract derives
            derives = extract_derives(lines, i)
            
            # Check for required traits
            has_clone = 'Clone' in derives or has_manual_impl(lines, struct_name, 'Clone')
            has_display = 'Display' in derives or has_manual_impl(lines, struct_name, 'Display')
            has_debug = 'Debug' in derives or has_manual_impl(lines, struct_name, 'Debug')
            has_eq = 'Eq' in derives or has_manual_impl(lines, struct_name, 'Eq')
            
            missing = []
            if not has_clone:
                missing.append('Clone')
            if not has_display:
                missing.append('Display')
            if not has_debug:
                missing.append('Debug')
            if not has_eq:
                missing.append('Eq')
            
            if missing:
                non_stt_structs.append({
                    'name': struct_name,
                    'line': i + 1,
                    'derives': derives,
                    'missing': missing
                })
    
    return non_stt_structs


def main():
    parser = create_review_parser(
        description="Detect data structures that don't satisfy StT (Eq + Clone + Display + Debug)"
    )
    args = parser.parse_args()
    context = ReviewContext(args)

    # Only check src/ files
    src_dir = context.repo_root / 'src'
    if not src_dir.exists():
        print("✗ No src/ directory found")
        return 1
    
    files = list(src_dir.rglob('*.rs'))
    print(f"Analyzing {len(files)} source files for StT compliance...")
    print("=" * 80)
    print()
    print("StT requirements: Eq + Clone + Display + Debug + Sized")
    print()
    
    all_violations = []
    
    for filepath in sorted(files):
        violations = analyze_file(filepath, context)
        if violations:
            all_violations.append((filepath, violations))
    
    if not all_violations:
        print("\n✓ All public structs satisfy StT requirements!")
        return 0
    
    # Group by what's missing
    missing_clone = []
    missing_display = []
    missing_debug = []
    missing_eq = []
    
    for filepath, violations in all_violations:
        for v in violations:
            if 'Clone' in v['missing']:
                missing_clone.append((filepath, v))
            if 'Display' in v['missing']:
                missing_display.append((filepath, v))
            if 'Debug' in v['missing']:
                missing_debug.append((filepath, v))
            if 'Eq' in v['missing']:
                missing_eq.append((filepath, v))
    
    total_count = sum(len(v) for _, v in all_violations)
    
    print(f"✗ Found {total_count} struct(s) that don't satisfy StT:\n")
    
    print(f"Summary by missing trait:")
    print(f"  Missing Clone:   {len(missing_clone)}")
    print(f"  Missing Display: {len(missing_display)}")
    print(f"  Missing Debug:   {len(missing_debug)}")
    print(f"  Missing Eq:      {len(missing_eq)}")
    
    print(f"\n{'='*80}")
    print("Detailed list:\n")
    
    for filepath, violations in sorted(all_violations, key=lambda x: len(x[1]), reverse=True):
        rel_path = context.relative_path(filepath)
        print(f"{rel_path}:")
        for v in violations:
            missing_str = ', '.join(v['missing'])
            derives_str = ', '.join(sorted(v['derives'])) if v['derives'] else 'none'
            print(f"  Line {v['line']}: {v['name']}")
            print(f"    Has derives: {derives_str}")
            print(f"    Missing: {missing_str}")
        print()
    
    return 1


if __name__ == '__main__':
    sys.exit(main())

File: src/bin/review_stt_compliance.rs
//! Review: StT Compliance
//!
//! Detects public structs that don't satisfy StT requirements.
//!
//! StT (Single-Threaded Type) = Eq + Clone + Display + Debug + Sized
//!
//! Binary: rusticate-review-stt-compliance

use anyhow::Result;
use ra_ap_syntax::{ast::{self, AstNode, HasName, HasVisibility, HasAttrs}, SyntaxKind, SourceFile, Edition};
use rusticate::{StandardArgs, find_rust_files};
use std::collections::HashSet;
use std::path::Path;
use std::time::Instant;

#[derive(Debug, Clone)]
struct NonSttStruct {
    name: String,
    line: usize,
    derives: HashSet<String>,
    missing: Vec<String>,
}

fn extract_derives(struct_node: &ast::Struct) -> HashSet<String> {
    let mut derives = HashSet::new();
    
    for attr in struct_node.attrs() {
        let attr_text = attr.to_string();
        if attr_text.starts_with("#[derive(") {
            // Extract traits: #[derive(Debug, Clone, ...)]
            if let Some(start) = attr_text.find('(') {
                if let Some(end) = attr_text.rfind(')') {
                    let traits_str = &attr_text[start + 1..end];
                    for trait_name in traits_str.split(',') {
                        derives.insert(trait_name.trim().to_string());
                    }
                }
            }
        }
    }
    
    derives
}

fn has_manual_impl(root: &ra_ap_syntax::SyntaxNode, struct_name: &str, trait_name: &str) -> bool {
    // Look for: impl Trait for StructName
    for node in root.descendants() {
        if node.kind() == SyntaxKind::IMPL {
            if let Some(impl_node) = ast::Impl::cast(node) {
                // Check if it's for our struct
                if let Some(self_ty) = impl_node.self_ty() {
                    let self_ty_str = self_ty.to_string();
                    if !self_ty_str.contains(struct_name) {
                        continue;
                    }
                    
                    // Check if it implements the trait
                    if let Some(trait_ref) = impl_node.trait_() {
                        let trait_str = trait_ref.to_string();
                        if trait_str.contains(trait_name) {
                            return true;
                        }
                    }
                }
            }
        }
    }
    false
}

fn analyze_file(file_path: &Path) -> Vec<NonSttStruct> {
    let content = match std::fs::read_to_string(file_path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let parsed = SourceFile::parse(&content, Edition::Edition2021);
    let tree = parsed.tree();
    let root = tree.syntax();

    let mut non_stt_structs = Vec::new();

    for node in root.descendants() {
        if node.kind() == SyntaxKind::STRUCT {
            if let Some(struct_node) = ast::Struct::cast(node.clone()) {
                // Check if it's public
                let is_public = if let Some(vis) = struct_node.visibility() {
                    vis.to_string().contains("pub")
                } else {
                    false
                };
                
                if !is_public {
                    continue;
                }
                
                let struct_name = if let Some(name) = struct_node.name() {
                    name.to_string()
                } else {
                    continue;
                };
                
                // Get line number
                let line_num = content[..node.text_range().start().into()]
                    .chars()
                    .filter(|&c| c == '\n')
                    .count() + 1;
                
                // Extract derives
                let derives = extract_derives(&struct_node);
                
                // Check for required traits
                let has_clone = derives.contains("Clone") || has_manual_impl(root, &struct_name, "Clone");
                let has_display = derives.contains("Display") || has_manual_impl(root, &struct_name, "Display");
                let has_debug = derives.contains("Debug") || has_manual_impl(root, &struct_name, "Debug");
                let has_eq = derives.contains("Eq") || has_manual_impl(root, &struct_name, "Eq");
                
                let mut missing = Vec::new();
                if !has_clone {
                    missing.push("Clone".to_string());
                }
                if !has_display {
                    missing.push("Display".to_string());
                }
                if !has_debug {
                    missing.push("Debug".to_string());
                }
                if !has_eq {
                    missing.push("Eq".to_string());
                }
                
                if !missing.is_empty() {
                    non_stt_structs.push(NonSttStruct {
                        name: struct_name,
                        line: line_num,
                        derives,
                        missing,
                    });
                }
            }
        }
    }

    non_stt_structs
}

fn main() -> Result<()> {
    let start_time = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();

    // Only check src/ files
    let src_files: Vec<_> = find_rust_files(&args.paths)
        .into_iter()
        .filter(|p| p.starts_with(base_dir.join("src")))
        .collect();

    println!("Analyzing {} source files for StT compliance...", src_files.len());
    println!("{}", "=".repeat(80));
    println!();
    println!("StT requirements: Eq + Clone + Display + Debug + Sized");
    println!();

    let mut all_violations = Vec::new();

    for file_path in &src_files {
        let violations = analyze_file(file_path);
        if !violations.is_empty() {
            all_violations.push((file_path.clone(), violations));
        }
    }

    if all_violations.is_empty() {
        println!("\n✓ All public structs satisfy StT requirements!");
        let elapsed = start_time.elapsed();
        eprintln!("\nCompleted in {}ms", elapsed.as_millis());
        return Ok(());
    }

    // Group by what's missing
    let mut missing_clone = 0;
    let mut missing_display = 0;
    let mut missing_debug = 0;
    let mut missing_eq = 0;

    for (_, violations) in &all_violations {
        for v in violations {
            if v.missing.contains(&"Clone".to_string()) {
                missing_clone += 1;
            }
            if v.missing.contains(&"Display".to_string()) {
                missing_display += 1;
            }
            if v.missing.contains(&"Debug".to_string()) {
                missing_debug += 1;
            }
            if v.missing.contains(&"Eq".to_string()) {
                missing_eq += 1;
            }
        }
    }

    let total_count: usize = all_violations.iter().map(|(_, v)| v.len()).sum();

    println!("✗ Found {} struct(s) that don't satisfy StT:\n", total_count);

    println!("Summary by missing trait:");
    println!("  Missing Clone:   {}", missing_clone);
    println!("  Missing Display: {}", missing_display);
    println!("  Missing Debug:   {}", missing_debug);
    println!("  Missing Eq:      {}", missing_eq);

    println!("\n{}", "=".repeat(80));
    println!("Detailed list:\n");

    // Sort by number of violations (descending)
    let mut sorted_violations = all_violations;
    sorted_violations.sort_by(|a, b| b.1.len().cmp(&a.1.len()));

    for (file_path, violations) in &sorted_violations {
        let rel_path = file_path.strip_prefix(&base_dir).unwrap_or(file_path);
        println!("{}:", rel_path.display());
        for v in violations {
            let missing_str = v.missing.join(", ");
            let mut derives_vec: Vec<_> = v.derives.iter().cloned().collect();
            derives_vec.sort();
            let derives_str = if derives_vec.is_empty() {
                "none".to_string()
            } else {
                derives_vec.join(", ")
            };
            println!("  Line {}: {}", v.line, v.name);
            println!("    Has derives: {}", derives_str);
            println!("    Missing: {}", missing_str);
        }
        println!();
    }

    let elapsed = start_time.elapsed();
    eprintln!("Completed in {}ms", elapsed.as_millis());

    Ok(())
}

